msal/application.py

```diff
@@ -947,6 +952,19 @@ def get_accounts(self, username=None):
         # Even in the rare case when an RT is revoked and then removed,
         # acquire_token_silent() would then yield no result,
         # apps would fall back to other acquire methods. This is the standard pattern.
+        if _is_running_in_cloud_shell():
+            # In Cloud Shell, user already signed in with an account.
+            # We pretend we have that account, for acquire_token_silent() to work.
+            # Note: If user acquire_token_by_xyz() using that account in MSAL later,
+            # the get_accounts() would return multiple accounts to calling app.
```

Comment:
> and then the user (or the app) would need to explicitly choose one account to proceed with `acquire_token_silent()`

In an Azure CLI command, there is no chance for the user to choose which account to use, since Azure CLI commands are designed to be run in **automation environments with no human intervention**.

> (2) If the scopes are outside of the allow-list, the `acquire_token_silent()` would return nothing, and then the calling app would naturally fall back to `acquire_token_by_xyz()` which is a normal MSAL behavior.

Similar to my previous comment, if `acquire_token_silent()` returns nothing, Azure CLI will give a hard failure, leaving the user no chance to re-login.

Rule of thumb: All Azure CLI commands except `az login` can require user intervention.


### My thought

Perhaps the best way is not to provide the Cloud Shell account so aggressively by `get_accounts()`. That is, `get_accounts()` should behave exactly like it is run out of Cloud Shell. If one wants to use Cloud Shell account, simply use a place holder in `acquire_token_silent()`: `acquire_token_silent(CLOUD_SHELL_ACCOUNT/MANAGED_IDENTITY_ACCOUNT)` to save us from all the fuss about `get_accounts()`.

Reply:
> Azure CLI commands are designed to be run in automation environments with no human intervention. ... except az login can require user intervention.

Understood. When I said "(when) `acquire_token_silent()` would return nothing, and then the calling app would naturally fall back to `acquire_token_by_xyz()`", I was not suggesting each az command would become interactive. I was merely saying "when `az foo bar` fails by token acquisition, az would prompt user to manually run `az login`", which is what az does today. We are on the same page here.

> provide the Cloud Shell account so aggressively by `get_accounts()`

I think I understand why Azure CLI does not feel such a need. It occurs to me that Azure CLI behaves like a "single account app", meaning, only one account can sign in at a time, and a user can only switch their different accounts by a real re-login. Therefore, Azure CLI keeps track of the current signed-in username and then only call `get_accounts(username="johndoe@contoso.com")` (and also expects only one account should be returned by that call). With this understanding, I do think I would like to change this PR to remove this behavior "The get_accounts(username="...") will return TWO accounts: the "Current Cloud Shell User" and a real johndoe@contoso.com." I'm considering changing it to "`get_accounts(username="johndoe@contoso.com")` will give you one AAD account, and `get_accounts(username="Current User")` will give you one Cloud Shell account.

On the other hand, MSAL Python's `get_accounts()` without the username parameter, is designed to facilitate "multiple account application (MAA)", which allows multiple accounts being signed in at the same time, and the app can switch among them without re-login. Please refer to [this doc](https://docs.microsoft.com/en-us/azure/active-directory/develop/single-multi-account), even though it is written for MSAL Android, but the underlying MAA concept is still the same. That is why we feel aggressively providing the Cloud Shell account in `get_accounts()` would allow an MAA to be run inside Cloud Shell *without* any code changes.

> If one wants to use Cloud Shell account, simply use a place holder in `acquire_token_silent()`: `acquire_token_silent(CLOUD_SHELL_ACCOUNT)` to save us from all the fuss about `get_accounts()`.

With the new proposal above, your Single-Account-Application could achieve same effect by this pattern:

```python
current_username = read_from_azcli_current_session() or msal.CurrentUser
    # Note: The msal.CurrentUser will be the new concept here,
    # the rest of the calling pattern is nothing new.
accounts = get_accounts(username=current_username)
if accounts:
    assert len(accounts) == 1, "MSAL usually returns only one account, especially for msal.CurrentUser"
    acquire_token_silent(scopes, accounts[0])
```
The alternative you proposed, will not be as simple as it seems.

```python
if is_running_inside_cloud_shell():
    acquire_token_silent(msal.CLOUD_SHELL_ACCOUNT)  # This seemingly-simple line would better only exist
                                                    # inside a CloudShell-only code path,
                                                    # which does not exist in all existing MSAL-powered apps.
                                                    # Adoption would be slow.
else:
    accounts = get_accounts(username="johndoe@contoso.com")
    if accounts:
       acquire_token_silent(scopes, accounts[0])
```

Reply:
As discussed offline, let's start with the first approach.